src/wayland/xdg_toplevel_tag.rs

```diff
@@ -0,0 +1,195 @@
+    /// so it's preferable for it to be human readable,
+    /// but it must be suitable for configuration files and should not be translated.
+    ///
+    /// Example of a tag would be "main window", "settings", "e-mail composer".
+    pub fn tag(&self) -> Option<Arc<str>> {
+        self.tag.lock().unwrap().clone()
+    }
+
+    /// Get description of a toplevel.
+    ///
+    /// This description may be shown to the user in UI or read by a screen reader for accessibility purposes.
+    pub fn description(&self) -> Option<Arc<str>> {
+        self.description.lock().unwrap().clone()
+    }
+}
+
+/// Handler trait for xdg toplevel tag events.
+pub trait XdgToplevelTagHandler:
+    GlobalDispatch<XdgToplevelTagManagerV1, ()> + Dispatch<XdgToplevelTagManagerV1, ()> + 'static
+{
+    /// Toplevel tag was set/updated.
+    #[allow(unused)]
+    fn on_set_tag(&mut self, toplevel: XdgToplevel, tag: String) {}
+
+    /// Toplevel description was set/updated.
+    #[allow(unused)]
+    fn on_set_description(&mut self, toplevel: XdgToplevel, description: String) {}
+}
+
+/// Delegate type for handling xdg toplevel tag events.
+#[derive(Debug)]
+pub struct XdgToplevelTagManager {
+    global: GlobalId,
+}
+
+impl XdgToplevelTagManager {
+    /// Creates a new delegate type for handling xdg toplevel tag events.
+    pub fn new<D: XdgToplevelTagHandler>(display: &DisplayHandle) -> Self {
+        let global = display.create_global::<D, XdgToplevelTagManagerV1, _>(1, ());
+        XdgToplevelTagManager { global }
+    }
+
+    /// Returns the [XdgToplevelTagManagerV1] global id.
+    pub fn global(&self) -> GlobalId {
+        self.global.clone()
+    }
+}
+
+impl<D: XdgToplevelTagHandler> GlobalDispatch<XdgToplevelTagManagerV1, (), D> for XdgToplevelTagManager {
+    fn bind(
+        _state: &mut D,
+        _dh: &DisplayHandle,
+        _client: &Client,
+        resource: New<XdgToplevelTagManagerV1>,
+        _data: &(),
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        data_init.init(resource, ());
+    }
+}
+
+impl<D: XdgToplevelTagHandler> Dispatch<XdgToplevelTagManagerV1, (), D> for XdgToplevelTagManager {
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        _resource: &XdgToplevelTagManagerV1,
+        request: xdg_toplevel_tag_manager_v1::Request,
+        _data: &(),
+        _dh: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        use xdg_toplevel_tag_manager_v1::Request;
+
+        match request {
+            Request::SetToplevelTag { toplevel, tag } => {
+                let data = toplevel.data::<XdgShellSurfaceUserData>().unwrap();
+
+                compositor::with_states(&data.wl_surface, |states| {
+                    states
+                        .data_map
+                        .insert_if_missing_threadsafe(XdgToplevelTagSurfaceData::default);
+                    let data = states.data_map.get::<XdgToplevelTagSurfaceData>().unwrap();
```

Comment:
The `insert_if_missing_threadsafe` and `get` could be combined into a `get_or_insert_threadsafe`.

Reply:
Oh, that's a nice addition, replaced.